src/OrchardCore/OrchardCore/Shell/ShellHost.cs

```diff
@@ -465,6 +470,28 @@ private bool CanReleaseShell(ShellSettings settings)
             return settings.State != TenantState.Disabled || _shellContexts.TryGetValue(settings.Name, out var value) && value.ActiveScopes == 0;
         }
 
+        public async Task RemoveShellAsync(ShellSettings settings)
+        {
+            if (_shellContexts.TryRemove(settings.Name, out var context))
+            {
+                _runningShellTable.Remove(settings);
+                _shellSettings.Remove(settings.Name, out settings);
+                context.Release();
+            }
+
+            //using (var serviceScope = (await CreateShellContextAsync(settings)).ServiceProvider.CreateScope())
+            //{
+            //    var shellEventHandlers = serviceScope.ServiceProvider.GetServices<IShellEventHandler>();
+            //    await shellEventHandlers.InvokeAsync(x => x.Removing(settings), _logger);
+            //}
+
+            await _shellSettingsManager.DeleteSettingsAsync(settings);
+
+            var tenantFolder = Path.Combine(_shellOptions.Value.ShellsApplicationDataPath, _shellOptions.Value.ShellsContainerName, settings.Name);
+
+            Directory.Delete(tenantFolder, true);
```

Comment:
How about having an `IShellEventHandler` to be fired here like under https://github.com/OrchardCMS/OrchardCore/compare/dev...Lombiq:issue/NEST-252? So other features can then hook into it and do their own deletion logic.

Reply:
I agree with @Piedone here, I would say all this deletion probably needs to be handled via event handlers.

I suspect, for example, that lucene may need to close down, or release a file lock, in order for the delete to be successful (I maybe wrong there, would have to look at how it resets indexes currently) but it would seem likely that the App_Data folder may contain files which need to be managed by their modules, rather than just deleted per say (one might archive all the media for example, for when they want to undelete it ;)

Reply:
For info there is a file lock per index which is created when the index is updated (e.g. on publishing a related indexed item). These file locks are released when the tenant is released, which is the case when we want to reload it, or when we disable it.

For now I opted to disable a tenant before removing it, so that the tenant is no more served and we can still create internally an isolated shell context on it, e.g. with a minimum set of features or all installed features, this depending on the stuff to do.

So here if the tenant is disabled (so released), the Lucene index files are no more locked, so we can just delete them e.g. when we delete the tenant folder.